### PR TITLE
Connects to Mongo

### PR DESCRIPTION
**Pull request by ogeagla**
Trying to address: https://github.com/metabase/metabase/issues/9867

The underlying issue seems to be that the Mongo Java driver cannot connect to an Atlas cluster which has DNS+SRV enabled without using a connection string using `mongodb+srv://` 

When enabling SRV, connecting to Atlas now works.

DNS SRV requires hostname to be an FQDN.  Therefore we can't always try to use SRV, especially since some users probably aren't using SRV.  In order to not introduce a regression for existing working configurations, we provide a toggle to allow users to choose whether or not to use SRV when connecting.  

If SRV is selected but hostname is not an FQDN, an error message is displayed telling user about it.

TODO:
 - [x] re-include default options (`ssl`, addtl opts)
 - [x] fallback still uses `authdb` and credentials
 - [x] validate data from Mongo is imported with new connection type
 - [x] `use-srv` toggle in Mongo connection UI
 - [x] tests covering bug
 - [x] using SRV w/o FQDN shows user an error

### CODE REVIEW

File: modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj
Comment (camsaul) [inline modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj]: can we add a new test that covers the original bug? Even if we just make sure we produce the Mongo right connection string or object or whatever giving some of the failing inputs described in #9867

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Could you please write down in comments or docstrings all the reasons we are doing thing the way we're doing them. (Especially why we use `connect` vs. `connect-srv`, and what those mean). Basically the stuff you explained to me as we were talking about the problem on Slack. That way in 2 years from now someone doesn't come in and have no clue why things are like this, assume it's duplicated code, and tear it out, or one of the community members doesn't decide to try Clojure for the first time and end up pulling their hair out trying to debug the wrong stuff for their connection.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: IMO having big 8-arg functions like this is hard to work with,

why not just make it a map instead and make your argslists be

```clj
[{:keys [host port user authdb pass dbname ssl additional-options]}]
```

(this applies to both `connect` and `connect-srv`

**Comment by camsaul** [comment]
I think this looks good, it should be ready to merge once we get a test added and write down what's going on somewhere. 👍

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: In general we try to follow the Clojure Style Guide as much as possible, e.g.

https://github.com/bbatsov/clojure-style-guide#function-positional-parameter-limit

for not using too many positional params

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Docstrings in

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Fixed

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: You don't have to change this now, it's not important, but we've been aiming for a line length of around 120 characters because that seems to be what GitHub displays on-screen without cutting things off so it makes looking at PRs nicer. I don't know how you would change it in IntelliJ but if it's a configuration thing you can add to the project dir feel free to do so, we have the equivalent settings for Emacs here https://github.com/metabase/metabase/blob/master/.dir-locals.el#L28

Comment (camsaul) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Oh yeah, sorry about being nitpicky here, but for next time we usually try to indent docstrings 2 spaces as suggested by the Clojure Style Guide
https://guide.clojure.style/#docstring-indentation.

Not something I would consider to be a reason to hold off on merging, just keep it in mind for next time

**Comment by camsaul** [comment]
Thanks for the docstring/style tweaks, all we need now is a test for the new stuff.

Since we're not about to spin up a cluster for this I think just checking that we generate the correct connection URI will be sufficient for now.

I think a few small tests would be good here:
*  One that checks that `localhost` gets `.localdomain` appended to it 
*  One that checks that something like `my-host.my-domain` is left as-is
*  One or two that checks that a URI with `mongodb+srv://` vs `mongo://` is generated in the situations where we want each to be used

Maybe you can do `with-redefs` and redef `mg/connect` and `connect-via-uri` to no-op and then look at the connection URI it would have used. Hopefully these won't be too hard to write

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: I wrote this Mongo code like 5 years ago when I was still relatively new to writing Clojure full-time. So if you want to refactor it to make it clearer or better please do not hesitate to do so. Our general rule if you want to make something clearer or better do it.

I was thinking about how one would write a test for this stuff and I think it would really be helpful to break this out into a few smaller functions. Specifically, I would decouple the code that generates a MongoDB connection object with the code that actually uses that object to establish a connection.

I played with this for a minute and think something like this will serve us better: it will be 100% easier to test, easier to follow (IMO) and extensible if we ever need to support one of the other options as well (hopefully not):

https://gist.github.com/camsaul/81bc81338726d4e618b8a36ce7433ff9

Comment (camsaul) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: I guess we can't comment on a Gist but my thinking was move the code that takes the Database `details` map and converts that to the appropriate MongoDB function into a couple of functions:

*  `normalize-details` to add defaults, remove empty keys like `:pass` (where needed) and rename keys stuff as appropriate
*  `details->mongo-connection-info` which will take the details and then return a map with a `:type` key for the type of connection we will be using. Other stuff such as URI are included in the map.

The we can have a separate `connect` function that takes the `mongo-connection-info` map and actually makes the connection. We can make it a multimethod that dispatches off of `:type` because both implementations are supposed to be returning the same thing.

The new tests will be super simple, you can just test `details->mongo-connection-info` directly and make sure that given the correct input from Database `:details` we get the correct connection object types

Comment (camsaul) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Those are just suggestions, I'm just trying to think of ways to make your life easier writing the tests

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: https://guide.clojure.style/#lisp-case

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Nice, good suggestion, I'll re-organize

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Some of our really old code doesn't have this, so feel free to fix it when you come across it.

To i18n stuff you need to wrap it in `trs` (for system-level stuff, e.g. log messages) or `tru` (for user-facing messages). This would be `tru` because it would be displayed in the DB setup panel when you tried to connect a DB if it failed.

The cannonical way to import `tru` and `trs` is `(:require [metabase.util.i18n :refer [tru trs])`. Some older code is using a a puppetlabs namespace and you can change that when you see it

Caveats: 

*  the `tru` and `trs` macros expand to some i18n record types we're using and you can't pass them directly to Java Exception constructors. So you have to wrap them in a call to `(str)`. That should probably just be part of the macro anyway so you don't have to remember it, but the guy who wrote didn't do it that way because it made it hard to do runtime translation of our Settings descriptions. But that can be fixed
*  it all has to be one line so you because of the way the lib we're using works. (I don't know why this is the case. I am planning on forking that lib and fixing that and my other complaints about it when I get a chance.)

Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Done

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Fixed

File: modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj
Comment (ogeagla) [reply modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj]: Done

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: https://guide.clojure.style/#body-indentation

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Sorry to be nitpicky but could you update the docstrings to reflect what's actually happening? This functon is not actually connecting to anything anymore so it's confusing to still have docstrings that say the function connects to something.

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: I don't know why the old code was indented that way but the same rule https://guide.clojure.style/#body-indentation applies here. The body of `finally` should be indented 2 spaces (hopefully your editor does this for you automatically)

File: modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj
Comment (camsaul) [inline modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj]: 👍 thanks for fixing this

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: doh, having trouble getting intellij to use .editorconfig though in theory should work

Comment (camsaul) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Sorry. Ignore the last comment. It's not the body of a `finally` form that is misindented but rather the body of the `ssh/with-ssh-tunnel` macro. The same https://guide.clojure.style/#body-indentation rule still applies

**Comment by camsaul** [comment]
Tests look good once you fix the indentation this is g2g

File: modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj
Comment (camsaul) [inline modules/drivers/mongo/test/metabase/driver/mongo/util_test.clj]: tests look good

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Fixed, good catch

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Fixed

**Review by camsaul — APPROVED** [review]

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (camsaul) [inline modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: two or more periods now

**Review by camsaul — APPROVED** [review]
I think this looks good, run thru it in the UI real quick to make sure the new option looks good and then we can merge

File: modules/drivers/mongo/src/metabase/driver/mongo/util.clj
Comment (ogeagla) [reply modules/drivers/mongo/src/metabase/driver/mongo/util.clj]: Fixed

**Pull request merged**
